File: leetcode/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.ret = []

    def dfs(self, node, path, remain):
        if node is None: return
        remain -= node.val
        path.append(node.val)
        if node.left is None and node.right is None and 0 == remain:
            self.ret += [path[:]]
        else:
            self.dfs(node.left, path, remain)
            self.dfs(node.right, path, remain)
        path.pop()

    def pathSum(self, root, sum_):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.dfs(root, [], sum_)
        return self.ret

File: leetcode/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
 public:
  void dfs(TreeNode* node, vector<int>& path, int remain)
  {
    path.push_back(node->val);
    remain -= node->val;

    if (!node->left && !node->right && !remain) {
      ret.push_back(path);
    } else {
      if (node->left) dfs(node->left, path, remain);
      if (node->right) dfs(node->right, path, remain);
    }
    path.pop_back();
  }

  vector<vector<int> > pathSum(TreeNode* root, int sum)
  {
    if (!root) return ret;
    vector<int> path;
    dfs(root, path, sum);
    return ret;
  }

 private:
  vector<vector<int> > ret;
};
